### PR TITLE
Add Functional Programming - Generics as Type Classes

### PR DESCRIPTION
**Pull request by jhwgh1968**
I put this under functional languages, since it is really an example of the concept of "type classes" in Rust. If this specific example is more of a design pattern, I would be happy to move it.

If no one beats me to it, I should be able to fix the two FIXMEs at the bottom sometime next week.

### CODE REVIEW

**Comment by jhwgh1968** [comment]
Hmm, the errors seem to be caused by broken links. I thought relative links would work in mdbook.

Any guidance on how to fix those?

File: functional/stateful-types.md [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+## Disadvantages
+
+This is a lot of typing. Depending on the amount of change caused by state transitions, an `InvalidState` enum value in an error type might be simpler.
+
+## Alternatives
+
+There are a number of simpler state machines, however, that have their own patterns:
+
+1. If the state transition is during construction/finalizing of an object, see [Builder Pattern](../creational/builder.md).
+1. If the state transitions don't change invariants much, see [Strategy Pattern](../behavioural/strategy.md).
```
Comment (simonsan) [inline functional/stateful-types.md:221]:
```suggestion
1. If the state transition is during construction/finalizing of an object, see [Builder Pattern](../patterns/creational/builder.md).
1. If the state transitions don't change invariants much, see [Strategy Pattern](../patterns/behavioural/strategy.md).
```

**Review by simonsan — COMMENTED** [review]
I think it's just that the `patterns` subdirectory was missing

**Comment by simonsan** [comment]
This is related to #100 isn't it? Maybe you can take the links from there, I posted some articles in it, that describe the pattern.

**Comment by jhwgh1968** [comment]
Thanks, @simonsan. I actually missed that PR somehow.

I have added the bottom of the two links you mentioned. Now my question is, how do I break up the link for the hard wrap limit? It is longer than 80 characters by itself.

File: functional/stateful-types.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+This is a lot of typing. Depending on the amount of change caused by state
+transitions, an `InvalidState` enum value in an error type might be simpler.
+
+## Alternatives
+
+There are a number of simpler state machines, however, that have their own patterns:
+
+1. If the state transition is during construction/finalizing of an object, see [Builder Pattern](../patterns/creational/builder.md).
+1. If the state transitions don't change invariants much, see [Strategy Pattern](../patterns/behavioural/strategy.md).
```
Comment (simonsan) [inline functional/stateful-types.md:241]:
```suggestion
1. If the state transition is during construction/finalizing of an object, see [Builder Pattern](../patterns/creational/builder.md).
2. If the state transitions don't change invariants much, see [Strategy Pattern](../patterns/behavioural/strategy.md).
```

File: functional/stateful-types.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+## Disadvantages
+
+This is a lot of typing. Depending on the amount of change caused by state
+transitions, an `InvalidState` enum value in an error type might be simpler.
+
+## Alternatives
+
+There are a number of simpler state machines, however, that have their own patterns:
+
+1. If the state transition is during construction/finalizing of an object, see [Builder Pattern](../patterns/creational/builder.md).
```
Comment (simonsan) [inline functional/stateful-types.md:240]:
```suggestion
1. If the state transition is during construction/finalizing of an object, see
[Builder Pattern](../patterns/creational/builder.md).
```

File: functional/stateful-types.md [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+This is a lot of typing. Depending on the amount of change caused by state
+transitions, an `InvalidState` enum value in an error type might be simpler.
+
+## Alternatives
+
+There are a number of simpler state machines, however, that have their own patterns:
+
+1. If the state transition is during construction/finalizing of an object, see
+[Builder Pattern](../patterns/creational/builder.md).
+1. If the state transitions don't change invariants much, see [Strategy Pattern](../patterns/behavioural/strategy.md).
```
Comment (simonsan) [inline functional/stateful-types.md:242]:
```suggestion
2. If the state transitions don't change invariants much, see [Strategy Pattern](../patterns/behavioural/strategy.md).
```

File: functional/stateful-types.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+There are a number of simpler state machines, however, that have their own patterns:
+
+1. If the state transition is during construction/finalizing of an object, see
+[Builder Pattern](../patterns/creational/builder.md).
+2. If the state transitions don't change invariants much, see [Strategy Pattern](../patterns/behavioural/strategy.md).
+
+## See also
+
+* [The Case for the Type State Pattern](https://www.novatec-gmbh.de/en/blog/the-case-for-the-typestate-pattern-the-typestate-pattern-itself/)
```
Comment (simonsan) [inline functional/stateful-types.md:246]:
```suggestion
* [The Case for the Type State Pattern](https://web.archive.org/web/20210325065112/https://www.novatec-gmbh.de/en/blog/the-case-for-the-typestate-pattern-the-typestate-pattern-itself/)
```

**Comment by simonsan** [comment]
Maybe this one is also a nice addition to `Read More`: 
https://web.archive.org/web/20210328164854/https://rustype.github.io/notes/notes/rust-typestate-series/rust-typestate-index

**Comment by jhwgh1968** [comment]
The second link makes sense to me, @simonsan, and seem to cover the type theory in more depth. I also put in a small wording change to be more precise, and helpful if people Google this.

I think it's a solid first version that is ready to merge, unless you have other additions.

**Comment by simonsan** [comment]
> 
> 
> The second link makes sense to me, @simonsan, and seem to cover the type theory in more depth. I also put in a small wording change to be more precise, and helpful if people Google this.
> 
> I think it's a solid first version that is ready to merge, unless you have other additions.

I need to give it a full read still, haven't completely reviewed it due to time constrains and just helped you fix the CI tests. Will take care of it the next two days. Cheers.

File: functional/stateful-types.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+1. If the state transition is during construction/finalizing of an object, see
+[Builder Pattern](../patterns/creational/builder.md).
+2. If the state transitions don't change invariants much, see [Strategy Pattern](../patterns/behavioural/strategy.md).
+
+## See also
+
+* [The Case for the Type State Pattern](https://web.archive.org/web/20210325065112/https://www.novatec-gmbh.de/en/blog/the-case-for-the-typestate-pattern-the-typestate-pattern-itself/)
+* [Rusty Typestate Series (an extensive thesis)](https://web.archive.org/web/20210328164854/https://rustype.github.io/notes/notes/rust-typestate-series/rust-typestate-index)
+can't find it again.
```
Comment (simonsan) [inline functional/stateful-types.md:249]:
```suggestion
```

**Review by simonsan — APPROVED** [review]
@MarcoIeni @pickfire Want to give it a read as well?

File: functional/stateful-types.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+```rust,ignore
+// this is a module to prevent users outside this crate from doing their own impls
+mod state_trait {
+    pub(crate) trait State {}
+
+    pub(crate) struct Init {
+        params: Vec<Param>,
+    }
+    impl State for Init {}
```
[the remark this answers is not in the document]
Comment (MarcoIeni) [reply functional/stateful-types.md:134]: maybe because of the comment at line 127?

**Comment by MarcoIeni** [comment]
What's the advantage of using this approach for state machines with respect to the one covered by the [rust book](https://doc.rust-lang.org/book/ch17-03-oo-design-patterns.html#implementing-transitions-as-transformations-into-different-types) ?

File: functional/stateful-types.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+order in the logic itself.
+
+In Rust, this is actually possible! The solution is to *change the type* in
+order to enforce the invariants.
+
+How? With a private generic parameter.
+
+Here is what that looks like:
+
+```rust,ignore
```
[the remark this answers is not in the document]
Comment (simonsan) [reply functional/stateful-types.md:126]: Could be nice to have a working example here, indeed. So people can copy and paste it and try it out/play around with it.

[the remark this answers is not in the document]
Comment (jhwgh1968) [reply functional/stateful-types.md:126]: I only did that to avoid repeating all the typing. I can easily put that back in and make it work.

**Comment by jhwgh1968** [comment]
The main advantage over the book, @MarcoIeni, is that this allows for shared data and IMO more readable shared impls.

I will type up that explanation in the PR.

File: functional/stateful-types.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (jhwgh1968) [reply functional/stateful-types.md:126]: Hmm, this is turning out to be trickier than I thought. I may have to refactor the example to make it simple enough. That will take me a bit.

File: functional/stateful-types.md [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+In Rust, the generic type parameter creates what is known as a "type class
+constraint", and each value used by an end user *actually changes the type of
+each instatiation*. In other words, `Vec<usize>` and `Vec<char>` *are two
+different types*.
+
+This is why `impl` blocks must specify generic parameters: different ones can
+have different `impl` blocks on them.
+
+It is recommended in rust to use generic types to enforce invariants. The best
```
Comment (simonsan) [inline functional/stateful-types.md:21]:
```suggestion
It is recommended in Rust to use generic types to enforce invariants. The best
```

File: functional/stateful-types.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+1. `set_param` can only be called in an "initial" state.
+2. `init` must be called before any scripts are compiled.
+3. `exec` can only be called in a "loaded" or "ready" state.
+4. `init` must be called *exactly once*.
+
+It would be possible to add these to the `Error` types instead of panicking.
+However, this solution is suboptimal. Not only would users of the code
+correctly have to `unwrap()` a result all the time, but the requirement to call
+init exactly once is still unenforcable across an entire program without
```
Comment (simonsan) [inline functional/stateful-types.md:112]:
```suggestion
`init()` exactly once is still unenforceable across an entire program without
```

File: functional/stateful-types.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (simonsan) [reply functional/stateful-types.md:126]: If you want to you can do that in another PR as well. It would be nice to have a working example that is checked, but not always necessary. Also this is quite complex so I'm fine with just refactoring it at a later stage, maybe even by someone else. @MarcoIeni thoughts?

File: functional/stateful-types.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+This sort of type transformation to represent state is also described in [the
+Rust Book](https://doc.rust-lang.org/book/ch17-03-oo-design-patterns.html#implementing-transitions-as-transformations-into-different-types),
+except the types are not generic. Using generic types has a key advantage: code
+re-use.
+
+First, it allows fields that are common to multiple states to be de-duplicated.
+That is, if the interpreter had a common field across the `Init`, `Loaded`, and
+`Ready` states, making them three structs would require the field to be
+consistent three times. By making the non-shared fields a generic, they are
```
Comment (simonsan) [inline functional/stateful-types.md:244]:
```suggestion
consistent three times. By making the non-shared fields generic, they are
```

**Review by simonsan — CHANGES_REQUESTED** [review]

File: functional/stateful-types.md [hunk 12]
```diff
@@ -0,0 +1,269 @@
+# Stateful Types: Generics as Type Classes
+
+## Description
+
+Rust's functional roots allow for it to be more expressive in the type system
```
Comment (pickfire) [inline functional/stateful-types.md:5]: This seems weird.

> Rust's functional roots allows it to be ...

Will this be better? I still find that weird though.

File: functional/stateful-types.md [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+"static typing" problems. A key part of this idea is the way generic types work.
+
+In C++ and Java, for example, generic types are a meta-programming construct for
+the compiler. `Vec<int>` and `Vec<char>` in C++ are just two different copies of
+the same boilerplate code for a `Vec` type, with two different types filled in.
+
+In Rust, the generic type parameter creates what is known as a "type class
+constraint", and each value used by an end user *actually changes the type of
+each instatiation*. In other words, `Vec<usize>` and `Vec<char>` *are two
+different types*.
```
Comment (pickfire) [inline functional/stateful-types.md:16]: Isn't this known as **monomorphization**?

File: functional/stateful-types.md [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+fn main() {
+    let mut interp = Interpreter::default();
+    interp.set_param(Param::GCSize(1024 * 1024 * 1024));
+    interp.init();
+    interp.compile_script("print('2 + 2')").unwrap();
+    interp.exec().unwrap();
+}
+```
```
Comment (pickfire) [inline functional/stateful-types.md:100]: Example seemed complicated. If someone like me who had no experience with state machine, this will be hard to understand. Just wondering could there be a simpler example?

Say, empty letter -> written letter -> closed letter?

File: functional/stateful-types.md [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+    pub(crate) struct Ready(Vec<CompiledScript>);
+    impl State for Ready {}
+}
+use state_trait::*;
+
+
+struct Interpreter<S: State> {
+    // same fields for execution, but now add:
+    state_data: S,
+}
```
Comment (pickfire) [inline functional/stateful-types.md:150]: Seemed extra to me. Why not just `Interpreter` -> `LoadedInterpreter` -> `ReadiedInterpreter` or something similar since they won't share common stuff?

File: functional/stateful-types.md [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+    }
+}
+```
+
+With this approach, if the user were to make a mistake and set a parameter
+after init:
+
+```rust,ignore
+fn main() {
+    let mut interp = Interpreter::<Init>::default().init();
```
Comment (pickfire) [inline functional/stateful-types.md:226]: Having this turbofish is the downside of what I mentioned. Not ergonomics and what if user does `Interpreter::<Loaded>::default()` instead?

**Comment by pickfire** [comment]
@jhwgh1968 Thanks for the PR and sorry for the wait. Although the example is simple, I believe it could be easier for someone with less computer science foundation to learn without having to know what is an interpreter. I gave an example (taken from somewhere else) but another other example with states could work.

Also, I discussed another way to do this typestate pattern without using generic which could provide both a more ergonomic API and easier to understand in the docs. I forgot the link to the original article but it was a good read.

File: functional/stateful-types.md (hunk 10 above)
Comment (pickfire) [reply functional/stateful-types.md:112]: Should we enable spell checkers?

File: functional/stateful-types.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (MarcoIeni) [reply functional/stateful-types.md:126]: It's up to jhwgh. If you want you can do it in another PR :)

File: functional/stateful-types.md (hunk 10 above)
Comment (MarcoIeni) [reply functional/stateful-types.md:112]: pickfire, you could open an issue about this. It doesn't seem like a bad idea. But we will need to have our own dictionary. By the way, let's continue the discussion in a proper issue, not here.

File: functional/stateful-types.md (hunk 13 above)
Comment (jhwgh1968) [reply functional/stateful-types.md:16]: After thinking about this, I could certainly add that. I was about to ask what C++ calls their template generation -- because I thought I heard a similar word -- but can't find it.

File: functional/stateful-types.md (hunk 12 above)
Comment (jhwgh1968) [reply functional/stateful-types.md:5]: I have rewritten this sentence completely.

File: functional/stateful-types.md (hunk 16 above)
Comment (jhwgh1968) [reply functional/stateful-types.md:229]: The default only being in the init state is precisely what I intended, for that exact reason.

I agree it's clunky, but also I think it's reasonable to recomment. If you don't put the turbofish in at all, I think the current stable has a good suggestion to lead you in the right direction.

**Comment by jhwgh1968** [comment]
Thanks for the comments, @pickfire.

This example is a little clunky, I agree. But unfortunately, all the times I have used it have been pretty specialized.

The latest is [some work I'm doing for the warc crate](https://github.com/jhwgh1968/warc/commit/2e38cf951ea5520cd4da9dc317cbe8f727201944). In that project, the generics are used to keep a common API for a `Record`, making the generic variants the way the body data of the record is accessed, based on how the record was created.

That is the uniquely Rust (or Haskell) way to approach the problem. It keeps the common parts of the API in one place, and surfaces any API misuse of the body accessors at compile time.

Those common parts are really key to the example, and they turned out to be missing when I actually tried to make it start working. Examples of that are extra hard to pick.

I will see if I can come up with a better example over the next several days, and report back.

**Comment by jhwgh1968** [comment]
I had a brain wave this evening: type states are not *really* what I'm trying to demonstrate. It's type classes using generics.

Given that, I came up with an example much similar to the times I have reached for this feature which is completely different. Hopefully it will look better now, and make more sense!

File: functional/generics-type-classes.md [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+the way monomorphization is implemented in the future.
+
+## Alternatives
+
+If a type seems to need a "split API" due to construction or partial
+initialization, consider the
+[Builder Pattern](../patterns/creational/builder.md) instead.
+
+If the split API is for methods in different state of a state machine, consider
+the [Strategy Pattern](../patterns/behavioural/strategy.md) instead.
```
Comment (pickfire) [inline functional/generics-type-classes.md:208]: I think it would be good to list type state as an alternative, either here or there we should have a comparison to let others know the differences between these two.

File: functional/generics-type-classes.md [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+
+## See also
+
+FIXME: are these links still relevant? I suspect not.
```
Comment (pickfire) [inline functional/generics-type-classes.md:212]: Just wondering, what other popular crates in the wild uses this pattern? The example seemed good (and specific) but it seemed a bit like the strategy pattern now, I am a bit confused between the two.

**Comment by simonsan** [comment]
Reopened #100 due to change of scope of this PR

File: functional/generics-type-classes.md (hunk 18 above)
Comment (jhwgh1968) [reply functional/generics-type-classes.md:212]: From what I have read on the page, this is how I see the difference:

* The strategy pattern uses generics to achieve multiple implementations of *the same* API. It is analagous to class inheritance in OO languages.
* This design focuses on *extending the API* using the generic type. It is more like the runtime polymorphism in Javascript -- except statically checked.

I will improve the text to reflect this.

Comment (jhwgh1968) [reply functional/generics-type-classes.md:212]: > Just wondering, what other popular crates in the wild uses this pattern?

Aside from my Warc example, I know I have seen some. I will find a few examples.

**Comment by jhwgh1968** [comment]
Alright, I have improved this significantly, and added examples.

Most important, I feel I have finally narrowed in on what is most relevant: the functional languages concept that Rust supports, and many non-function programmers (including my past self) would benefit from when they are designing types.

**Comment by jhwgh1968** [comment]
The failure to build is:

```
  --> /tmp/mdbook-3OvrCB/functional/generics-type-classes.md:121:17
   |
17 |     use super::{bootp, nfs};
   |                 ^^^^^  ^^^ no `nfs` in the root
   |                 |
   |                 no `bootp` in the root

error: aborting due to previous error
```

I defined those modules in the example, so I presume this is something weird with resolving in rustdoc/mdbook. How do I fix this?

**Comment by MarcoIeni** [comment]
> I defined those modules in the example, so I presume this is something weird with resolving in rustdoc/mdbook. How do I fix this?

I don't know unfortunately. You can put the `ignore` flag, or you can open an issue in the mdbook repo to see if it is a bug or a known limitation.

File: functional/generics-type-classes.md [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+
+Rust's type system is designed more like functional languages (like Haskell)
+rather than imperative languages (like Java and C++). As a result, Rust can turn
+many kinds of programming problems into "static typing" problems. This is one
+of the biggest wins of choosing a functional language, and is critical to many
+of Rust's compile time guarantees.
+
+A key part of this idea is the way generic types work. In C++ and Java, for
+example, generic types are a meta-programming construct for the compiler.
+`Vec<int>` and `Vec<char>` in C++ are just two different copies of the same
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:13]:
```suggestion
`vector<int>` and `vector<char>` in C++ are just two different copies of the same
```
Let's use cpp names, right?

File: functional/generics-type-classes.md [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+boilerplate code for a `Vec` type (known as a `template`)  with two
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:14]:
```suggestion
`vector<int>` and `vector<char>` in C++ are just two different copies of the same
boilerplate code for a `vector` type (known as a `template`)  with two
```
Should we use cpp names?

File: functional/generics-type-classes.md [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+particular members of a type class, but extra behavior as well.
+
+The nearest equivalent is the runtime polymorphism in Javascript and Python,
+where new members can be added to objects willy-nilly by any constructor.
+Unlike those languages, however, all of Rust's additional methods can be type
+checked when they are used, because their generics are statically defined. That
+makes them more usable while remaining safe.
+
+An example follows.
+
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:39]:
```suggestion
```
do we need this?

File: functional/generics-type-classes.md [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+A key part of this idea is the way generic types work. In C++ and Java, for
+example, generic types are a meta-programming construct for the compiler.
+`Vec<int>` and `Vec<char>` in C++ are just two different copies of the same
+boilerplate code for a `Vec` type (known as a `template`)  with two
+different types filled in.
+
+In Rust, a generic type parameter creates what is known in functional
+languages as a "type class constraint", and each different parameter filled in
+by an end user *actually changes the type*. In other words, `Vec<usize>` and
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:19]:
```suggestion
by an end user *actually changes the type*. In other words, `Vec<isize>` and
```
`isize` is the corresponding of cpp `int`

File: functional/generics-type-classes.md [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+The way the current struct is designed leaves the protocol decision until
+runtime. That means any method that applies to one protocol and not the other
+requires the programmer to do a runtime check.
+
+Here is how getting an NFS mount point would look:
+
+```rust,ignore
+struct FileDownloadRequest {
+    file_name: PathBuf,
+    authentication: Either<crate::nfs::AuthInfo, crate::bootp::AuthInfo>,
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:77]: `Either` is not a rust concept. This concept may not be familiar to a reader that comes from imperative languages. Or which approaches rust as their first programming language.
Should we have something like:

```
enum AuthInfo {
    Nfs(crate::nfs::AuthInfo),
    Bootp(crate::bootp::AuthInfo),
}
```
?

**Comment by MarcoIeni** [comment]
Also please, do not force push. It's easier for us to see what changes from review to review.
Everything will be squashed in one commit anyway when we merge this.

File: functional/generics-type-classes.md [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+
+    /// Gets an NFS mount point if this is an NFS request. Otherwise,
+    /// return None.
+    pub fn mount_point(&self) -> Option<&Path> {
+        self.mount_point.as_ref()
+    }
+}
+```
+
+Every caller of `get_mount point` must check for `None` and write code to handle
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:92]:
```suggestion
Every caller of `mount_point()` must check for `None` and write code to handle
```

File: functional/generics-type-classes.md [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+        }
+    }
+    pub struct Bootp(); // no additional metadata
+    impl ProtoKind for Bootp {
+        type AuthInfo = bootp::AuthInfo;
+        fn auth_info(&self) -> Self::AuthInfo {
+            bootp::AuthInfo()
+        }
+    }
+}
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:145]: can you add new lines to separate structs and impl blocks?

File: functional/generics-type-classes.md [hunk 26]
```diff
@@ -0,0 +1,268 @@
...
+
+// private module, lest outside users invent their own protocol kinds!
+mod proto_trait {
+    use std::path::{Path, PathBuf};
+    use super::{bootp, nfs};
+    pub(crate) trait ProtoKind {
+        type AuthInfo;
+        fn auth_info(&self) -> Self::AuthInfo;
+    }
+    pub struct Nfs(nfs::AuthInfo, PathBuf); // the mount point metadata
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:126]: it's not clear that the comment is referred to the `PathBuf` element.
Maybe it's better have named struct fields?

**Comment by MarcoIeni** [comment]
image
Can you specify that with code size we mean the size of the binary? Feel free to rephrase this as you prefer of course.

File: functional/generics-type-classes.md [hunk 27]
```diff
@@ -0,0 +1,268 @@
...
+the way monomorphization is implemented in the future.
+
+## Alternatives
+
+If a type seems to need a "split API" due to construction or partial
+initialization, consider the
+[Builder Pattern](../patterns/creational/builder.md) instead.
+
+If the API between types does not change -- only the behavior does -- then the
+[Strategy Pattern](../patterns/behavioural/strategy.md) is better used instead.
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:217]:
```suggestion
* If a type seems to need a "split API" due to construction or partial
  initialization, consider the
  [Builder Pattern](../patterns/creational/builder.md) instead.

* If the API between types does not change -- only the behavior does -- then the
  [Strategy Pattern](../patterns/behavioural/strategy.md) is better used instead.
```

File: functional/generics-type-classes.md [hunk 28]
```diff
@@ -0,0 +1,268 @@
...
+
+[^4] Example:
+[https://docs.rs/stm32f30x-hal/0.1.0/stm32f30x_hal/gpio/gpioa/struct.PA0.html](
+https://docs.rs/stm32f30x-hal/0.1.0/stm32f30x_hal/gpio/gpioa/struct.PA0.html)
+
+[^5] See:
+[https://docs.rs/hyper/0.14.5/hyper/client/struct.Client.html](
+https://docs.rs/hyper/0.14.5/hyper/client/struct.Client.html)
+
+[^6] See:
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:263]:
```suggestion
[^1]: See: [impl From\<CString\> for Vec\<u8\>](
https://doc.rust-lang.org/stable/src/std/ffi/c_str.rs.html#799-801)


[^2]: See: [impl\<T\> From\<Vec\<T, Global\>\> for BinaryHeap\<T\>](
https://doc.rust-lang.org/stable/src/alloc/collections/binary_heap.rs.html#1345-1354)

[^3]: See: [impl\<'_\> ToString for Cow\<'_, str>](
https://doc.rust-lang.org/stable/src/alloc/string.rs.html#2235-2240)

[^4]: Example:
[https://docs.rs/stm32f30x-hal/0.1.0/stm32f30x_hal/gpio/gpioa/struct.PA0.html](
https://docs.rs/stm32f30x-hal/0.1.0/stm32f30x_hal/gpio/gpioa/struct.PA0.html)

[^5]: See:
[https://docs.rs/hyper/0.14.5/hyper/client/struct.Client.html](
https://docs.rs/hyper/0.14.5/hyper/client/struct.Client.html)

[^6]: See:
```
This should fix footnotes, see https://github.com/rust-lang/mdBook/blob/master/tests/dummy_book/src/first/markdown.md

**Comment by MarcoIeni** [comment]
This looks great. Thanks!
Other than the minor comments I left it looks good to me.

File: functional/generics-type-classes.md (hunk 20 above)
Comment (jhwgh1968) [reply functional/generics-type-classes.md:14]: Absolutely! Brain fade on my part :smile:

File: functional/generics-type-classes.md (hunk 23 above)
Comment (jhwgh1968) [reply functional/generics-type-classes.md:77]: `Either` is defined in [the `either` crate](http://docs.rs/either), which I think is easily discoverable. I will clarify that.

Comment (jhwgh1968) [reply functional/generics-type-classes.md:77]: Actually, the `enum` is better. I'll fix it.

**Comment by jhwgh1968** [comment]
I have a couple more review comments to go, but I am getting close.

> Can you specify that with code size we mean the size of the binary?

Done.

> Also please, do not force push. It's easier for us to see what changes from review to review.
Everything will be squashed in one commit anyway when we merge this.

On other PRs I saw maintainers doing "merge commits" with master, which usually creates a giant mess in git. I will keep that in mind that you always squash, and not worry about it.

> The problem is that mdbook automatically adds fn main() { } around everything because it doesn't exist.

Ah, that makes sense. I fix it.

**Comment by jhwgh1968** [comment]
The `main` function I added to the compiling example is currently empty. I am planning to put in a simple example of using the types, but I would have to think about it (because construction is something I skipped over).

**Comment by MarcoIeni** [comment]
> On other PRs I saw maintainers doing "merge commits" with master, which usually creates a giant mess in git. I will keep that in mind that you always squash, and not worry about it.

We set this in December: 
image

File: functional/generics-type-classes.md [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+
+    /// Gets an NFS mount point if this is an NFS request. Otherwise,
+    /// return None.
+    pub fn mount_point(&self) -> Option<&Path> {
+        self.mount_point.as_ref()
+    }
+}
+```
+
+Every caller of `mount point()` must check for `None` and write code to handle
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:96]:
```suggestion
Every caller of `mount_point()` must check for `None` and write code to handle
```

File: functional/generics-type-classes.md [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+Rust's type system is designed more like functional languages (like Haskell)
+rather than imperative languages (like Java and C++). As a result, Rust can turn
+many kinds of programming problems into "static typing" problems. This is one
+of the biggest wins of choosing a functional language, and is critical to many
+of Rust's compile time guarantees.
+
+A key part of this idea is the way generic types work. In C++ and Java, for
+example, generic types are a meta-programming construct for the compiler.
+`vector<int>` and `vector<char>` in C++ are just two different copies of the
+same boilerplate code for a `Vec` type (known as a `template`)  with two
```
Comment (MarcoIeni) [inline functional/generics-type-classes.md:14]:
```suggestion
same boilerplate code for a `vector` type (known as a `template`)  with two
```

**Comment by simonsan** [comment]
Cool! I'll also read through it again in the next days. <3

**Review by MarcoIeni — DISMISSED** [review]
This is great! I approved it. simonsan, merge this if it looks good to you once you have read it :)

**Comment by MarcoIeni** [comment]
The CI is failing for unrelated reasons:
image

**Comment by jhwgh1968** [comment]
Thanks, @MarcoIeni! I unfortunately could not think of an easy constructor for main, but I think this example is clear enough as it is.

File: functional/generics-type-classes.md [hunk 31]
```diff
@@ -0,0 +1,286 @@
...
+
+Second, it makes the `impl` blocks easier to read, because they are broken down
+by state. Methods common to all states are typed once in one block, and methods
+unique to one state are in a separate block.
+
+Both of these mean there are fewer lines of code, and they are better organized.
+
+## Disadvantages
+
+This current increases the size of the binary, due to the way monomorphization
```
Comment (simonsan) [inline functional/generics-type-classes.md:224]:
```suggestion
This currently increases the size of the binary, due to the way monomorphization
```

**Review by simonsan — APPROVED** [review]

**Comment by simonsan** [comment]
Nice, thank you! :)
